csrc/custom_all_reduce.cuh

```diff
@@ -131,15 +131,26 @@ DINLINE O downcast(array_t<float, O::size> val) {
 }
 
 static DINLINE void st_flag_release(FlagType* flag_addr, FlagType flag) {
+#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ >= 700
   asm volatile("st.release.sys.global.u32 [%1], %0;" ::"r"(flag),
                "l"(flag_addr));
+#else
+  asm volatile("membar.sys; st.volatile.global.u32 [%1], %0;" ::"r"(flag),
+               "l"(flag_addr));
+#endif
 }
 
 static DINLINE FlagType ld_flag_acquire(FlagType* flag_addr) {
   FlagType flag;
+#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ >= 700
   asm volatile("ld.acquire.sys.global.u32 %0, [%1];"
                : "=r"(flag)
                : "l"(flag_addr));
+#else
+  asm volatile("ld.volatile.global.u32 %0, [%1]; membar.gl;"
```

Comment:
In the NCCL polyfill there's a `.sys` scope here. Should we also include it here?

Reply:
I tried that first, but got an `error: Modifier '.volatile' cannot be combined with modifier '.sys'` ([build log](https://github.com/sasha0552/pascal-pkgs-ci/actions/runs/11015747881/job/30589559877#step:3:1300)). I wonder why this works for NCCL, but as I read in the documentation, `.volatile` implies `.sys`.

> ld.volatile has the same memory synchronization semantics as ld.relaxed.sys.

-- https://docs.nvidia.com/cuda/parallel-thread-execution/index.html